src/main/java/io/github/opencubicchunks/cubicchunks/core/server/chunkio/region/ShadowPagingRegion.java

```diff
@@ -74,25 +81,79 @@ private ShadowPagingRegion(FileChannel file, SectorTracker<K> sectorTracker, IHe
 		this.sectorTracker = sectorTracker;
 	}
 
-	@Override public synchronized void writeValue(K key, ByteBuffer value) throws IOException {
+	/**
+	 * Writes an entry without updating the region headers on disk.
+	 * @return whether or not {@link #file} needs to be flushed before the headers are be updated
+	 */
+	protected boolean writeValueFirst(K key, ByteBuffer value) throws IOException {
 		if (value == null) {
 			this.sectorTracker.removeKey(key);
-			updateHeaders(key);
+			return false; //the data isn't actually changed on-disk, so we don't need to flush the channel before updating the headers
+		} else {
+			int size = value.remaining();
+			int sizeWithSizeInfo = size + Integer.BYTES;
+			int numSectors = this.getSectorNumber(sizeWithSizeInfo);
+			// this may throw UnsupportedDataException if data is too big
+			RegionEntryLocation location = this.sectorTracker.reserveForKey(key, numSectors);
```

Comment:
This is still broken because reserveForKey marks the old entry as unused after new value for it is reserved. So it's going to be free to be reallocated for next entries in a batch write.